dask/dataframe/operation.py

```diff
@@ -0,0 +1,402 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, replace
+from functools import cached_property
+from typing import Any, Callable, Hashable
+
+from dask.base import tokenize
+from dask.highlevelgraph import HighLevelGraph
+from dask.operation import (
+    CollectionOperation,
+    FusableOperation,
+    FusedOperations,
+    LiteralInputs,
+    fuse_subgraph_callables,
+)
+from dask.optimization import SubgraphCallable
+from dask.utils import apply
+
+
+@dataclass(frozen=True)
+class FrameOperation(CollectionOperation):
+    """Abtract DataFrame-based CollectionOperation"""
+
+    @property
+    def meta(self) -> Any:
+        """Return DataFrame metadata"""
+        raise NotImplementedError
+
+    def replace_meta(self, value) -> CollectionOperation:
+        """Return a new operation with different meta"""
+        raise ValueError(f"meta cannot be modified for {type(self)}")
+
+    @property
+    def divisions(self) -> tuple | None:
+        """Return DataFrame divisions"""
+        raise NotImplementedError
+
+    def replace_divisions(self, value) -> CollectionOperation:
+        """Return a new operation with different divisions"""
+        raise ValueError(f"divisions cannot be modified for {type(self)}")
+
+    @property
+    def npartitions(self) -> int | None:
+        """Return partition count"""
+        if not self.divisions:
+            return None
+        return len(self.divisions) - 1
+
+    @property
+    def collection_keys(self) -> list:
+        """Return list of all collection keys"""
+        if self.npartitions is None:
+            raise ValueError
+        return [(self.name, i) for i in range(self.npartitions)]
+
+
+@dataclass(frozen=True)
+class CompatFrameOperation(FrameOperation):
+    """Pass-through FrameOperation
+
+    This class acts as a container for the name, meta,
+    divisions, and graph (HLG) of a "legacy" collection.
+    Note that a ``CompatFrameOperation`` may not have any
+    dependencies.
+    """
+
+    _dsk: dict | HighLevelGraph
+    _name: str
+    _meta: Any
+    _divisions: tuple | None
+    parent_meta: Any | None = None
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def meta(self) -> Any:
+        return self._meta
+
+    def replace_meta(self, value) -> CompatFrameOperation:
+        return replace(self, _meta=value)
+
+    @property
+    def divisions(self) -> tuple | None:
+        return tuple(self._divisions) if self._divisions is not None else None
+
+    def replace_divisions(self, value) -> CompatFrameOperation:
+        return replace(self, _divisions=value)
```

Comment:
I'm not convinced that this is any easier to use than just making `meta` and `divisions` proper fields on the `FrameOperation` class:

```python
@dataclass(frozen=True)
class FrameOperation(CollectionOperation):
    """Abtract DataFrame-based CollectionOperation"""
    meta: Any
    divisions: tuple[Any]

op = FrameOperation(meta=df, divisions=(None,)*npartitions)
op.reinitialize(meta=new_meta)
```

It forces the positioning of the args in the `__init__`, but perhaps it's not too bad for these cases?

Reply:
My interest in avoiding "field" status for meta and divisions is entirely because I imagine that we will eventually want to allow the Operation itself to figure out what the meta and divisions should be. If we add meta and divisions as fields, we are establishing that we typicall expect meta and divisions to be known before a FieldOperation is initialized. With that said, this assumptions will certainly be true at first, and I could be wrong about the benefits of breaking away from this behavior.